### DISCUSSION TITLE
Symfony 6 forms always cause empty form or "Field has already been rendered" expection

### DISCUSSION THREAD

**Discussion by typoworx-de** [symfony/symfony#48390] [Q&A] [answered]
I'm trying to implement a simple, stupid form as documented in symfony docs:

My Controller
```php
public function index (Request $request) : Response
{
        $form = $this->createFormBuilder($this->customer)
            ->add('firstname', TextareaType::class)
            ->add('submit', SubmitType::class)
            ->getForm()
        ;
        
        return $this->render(
            'frontend/controller/profile.html.twig', [
               'form' => $form
            ]
         );
}
```
Template (Twig)
```twig/html
        {{form_start(form, {'method': 'POST', 'role':'form' } )}}
            {{ form_errors(form) }}
            {{ form_widget(form) }}

             // Commenting this in causes Exception! Without empty <form />
            {# { form_row(form.firstname) } #}
        {{ form_end(form, {'render_rest': true}) }}
```

The result is always strange... without explicit calling twig form_row to manually a field, the form is rendered empty (only form-tag without any content inside). Trying to manually render it throws:

```
[BadMethodCallException][RuntimeError]
HTTP 500 Internal Server Error
An exception has been thrown during the rendering of a template ("Field "form" has already been rendered, save the result of previous render call to a variable and output that instead.").
```

I'm trying to get it work since at least 3 hours without result and this is totally confusing! If the field would have been rendered, why isn't it when trying to let it render?

**Comment by MatTheCat** [comment]
Your controller pass a `formProfile` variable to your template, but your template refers a `form` one; is that really the corresponding code?

Could you provide a [reproducer](https://symfony.com/doc/current/contributing/code/reproducer.html)?

**Reply by typoworx-de** [reply]
Yes.. you're right. I'm using formProfile and stripping down the example I've forgotten to do it with formProfile in the render arguments. But that's not the solution at all :-(

I fixed the example above (formProfile -> form for the argument-name for render.

**Reply by typoworx-de** [reply]
@MatTheCat the reproducer was a good idea! Surprisingly there's working and after some fiddling around installing more composer-packages that I'm using in the faulty project... I've stripped down the problem to this:

base.html.twig (simplified to essential code):
```html
...
<body>
        <!-- placing form here works! -->
        {{ form_start(form) }}
        {{ form_end(form) }}
        {% if block('main') is not empty %}
            <main>
               <!-- placing form here works NOT?! -->
               {{ form_start(form) }}
               {{ form_end(form) }}

                {% block main %}{% endblock %}
            </main>
        {% endif %}
</body>
...
``` 

My profile.html.twig looks like this, but I think the problem is earlier in the ```if block('main')``` part:

profile.html.twig:
```
{% block main %}
    {{ form_start(form) }}
    {{ form_end(form) }}
{% endblock %}
```

Any Idea why the form is only working in base.html.twig and only outside this block:
```
{% if block('main') is not empty %}
...
{% endif %}
```

**Reply by typoworx-de** [reply]
Oh damn... after reading the code twice... I smelled the culprit. Obviously the if block('main') will render the content first-time and second-time if it will be not empty :face_with_spiral_eyes:

**Reply by typoworx-de** [reply]
This is how I've improved the twig-block causing the problem. I've totally missed that indeed the if condition will already render the block and I've wondered how and where this should have been happened! Totally logic now where I've found out whats wrong.

My improved code for checking if a block exists being able to skip f.e. aside block:

```
        {% with { content: block('main') } %}
            {% if content is not empty %}
                <main>
                    {{ content|raw }}
                </main>
            {% endif %}
        {% endwith %}
```

**Reply by MatTheCat** [reply]
Does `defined` solve your problem? This is how the Twig documentation says to test if a block exist: https://twig.symfony.com/doc/3.x/functions/block.html

**Comment by typoworx-de** [comment]
Culprit found... indeed my fault, see above. Ticket can be closed here!

**Comment by typoworx-de** [comment, marked as answer]
For the problem see comment above: https://github.com/symfony/symfony/discussions/48390#discussioncomment-4266746
